experiment/simulation/js/flipflop.js

```diff
@@ -0,0 +1,780 @@
+        return gate.q;
+    }
+    else if (pos === "Q'") {
+        return gate.qbar;
+    }
+    // But if the gate is not a flipflop, but an input bit, then return the value of the input
+    else {
+        return gate.output
+    }
+}
+
+// done checking
+export function getResultJK(ff) {
+    // check if flipflop type is Gate object
+    if (ff.constructor.name === "Gate") {
+        return;
+    }
+
+
+    if (getOutputJK(ff.k[0], ff.k[1]) != null && getOutputJK(ff.j[0], ff.j[1]) != null && getOutputJK(ff.clk[0], ff.clk[1]) != null) {
+        ff.generateOutput();
+    }
+
+    return;
+}
+
+// done checking
+export function checkConnectionsJK() {
+    let correctConnection = true;
+    for (let ffID in flipFlops) {
+        const gate = flipFlops[ffID];
+        // For Full Adder objects
+        // Check if all the outputs are connected
+        const id = document.getElementById(gate.id);
+        if (!gate.qIsConnected) {
+            printErrors("Q of Flip-Flop must be connected\n",id);
+            correctConnection = false;
+            break;
+        }
+        // Check if all the inputs are connected
+        if (gate.k == null || gate.k.length === 0) {
+            printErrors("K of Flip-Flop must be connected\n",id);
+            correctConnection = false;
+            break;
+        }
+        if (gate.j == null || gate.j.length === 0) {
+            printErrors("J of Flip-Flop must be connected\n",id);
+            correctConnection = false;
+            break;
+        }
+        if (gate.clk == null || gate.clk.length === 0) {
+            printErrors("Clock of Flip-Flop must be connected\n",id);
+            correctConnection = false;
+            break;
+        }
+    }
+    let count = 0;
+    for (let gateId in gates) {
+        const gate = gates[gateId];
+        if(gate.type === "Input" && gate.output === true)
+        {
+            count++;
+        }
+        if (gate.isInput) {
+            if (!gate.isConnected) {
+                correctConnection = false;
+                break;
+            }
+        }
+        else if (gate.isOutput) {
+            if (gate.inputs.length === 0) {
+                correctConnection = false;
+                break;
+            }
+        }
+        else {
+            if (gate.inputPoints.length != gate.inputs.length) {
+                correctConnection = false;
+            }
+            else if (!gate.isConnected && !gate.isOutput) {
+                correctConnection = false;
+            }
+        }
+    }
+
+    if (correctConnection) {
+        if(count !==2)
+        {
+            printErrors("J,K must be initialised to 1\n",null);
+            
+            return false;
+        }
+        return true;
+    }
+    else {
+        
+        return false;
+    }
+}
+
+export function simulateFFJK() {
+    for (let ffID in flipFlops) {
+        const gate = flipFlops[ffID];
+        getResultJK(gate);
+    }
+}
+
+export function testSimulateFFJK(flipFlops) {

```

Comment:
Why do we need these 2 identical functions?

Reply:
The first function simulates the circuit with values on the screen while the other one simulates the circuit for the purpose of validation. The first function uses a global dictionary of flip-flops whereas the other one uses a parameter

Reply:
But they are essentially the same functions so why do we need to duplicate the code? If we want 2 separate names for better readability, we can just use 2 different names for the same function.

Reply:
The second function has a parameter, so it will create a new copy of flip-flops and wouldn't affect the original flip-flops array, this is done to not have any conflicts between submit and simulate